google/auth/_default.py

```diff
@@ -0,0 +1,295 @@
+    explicit_file = os.environ.get(_CREDENTIALS_ENV)
+    if explicit_file is not None:
+        return _load_credentials_from_file(os.environ[_CREDENTIALS_ENV])
+    else:
+        return None, None
+
+
+def _get_gcloud_sdk_project_id(config_path):
+    """Gets the project ID from the Cloud SDK's configuration.
+
+    Args:
+        config_path (str): The path to the Cloud SDK's config directory,
+            for example ``~/.config/gcloud``.
+
+    Returns:
+        Optional[str]: The project ID.
+    """
+    config_file = os.path.join(config_path, _CLOUDSDK_ACTIVE_CONFIG_FILENAME)
+
+    if not os.path.exists(config_file):
+        return None
+
+    config = configparser.RawConfigParser()
+
+    try:
+        config.read(config_file)
+    except configparser.Error:
+        return None
+
+    if config.has_section(_CLOUDSDK_PROJECT_CONFIG_SECTION):
+        return config.get(
+            _CLOUDSDK_PROJECT_CONFIG_SECTION, _CLOUDSDK_PROJECT_CONFIG_KEY)
+
+
+def _get_gcloud_sdk_config_path():
+    """Returns the absolute path the the Cloud SDK's configuration directory.
+
+    Returns:
+        str: The Cloud SDK config path.
+    """
+    if _CLOUDSDK_CONFIG_ENV in os.environ:
+        return os.environ[_CLOUDSDK_CONFIG_ENV]
+
+    # Non-windows systems store this at ~/.config/gcloud
+    if os.name != 'nt':
+        return os.path.join(
+            os.path.expanduser('~'), '.config', _CLOUDSDK_CONFIG_DIRECTORY)
+    # Windows systems store config at %APPDATA%\gcloud
+    else:
+        if 'APPDATA' in os.environ:
+            return os.path.join(
+                os.environ['APPDATA'], _CLOUDSDK_CONFIG_DIRECTORY)
+        else:
+            # This should never happen unless someone is really
+            # messing with things, but we'll cover the case anyway.
+            drive = os.environ.get('SystemDrive', 'C:')
+            return os.path.join(
+                drive, '\\', _CLOUDSDK_CONFIG_DIRECTORY)
+
+
+def _get_gcloud_sdk_credentials():
+    """Gets the credentials and project ID from the Cloud SDK."""
+    # Get the Cloud SDK's configuration path.
+    config_path = _get_gcloud_sdk_config_path()
+
+    # Check the config path for the credentials file.
+    credentials_filename = os.path.join(
+        config_path, _CLOUDSDK_CREDENTIALS_FILENAME)
+
+    if not os.path.exists(credentials_filename):
+        return None, None
+
+    credentials, project_id = _load_credentials_from_file(
+        credentials_filename)
+
+    if not project_id:
+        project_id = _get_gcloud_sdk_project_id(config_path)
+
+    return credentials, project_id
+
+
+def _get_gae_credentials():
+    """Gets Google App Engine App Identity credentials and project ID."""
+    return None, None
+
+
+def _get_gce_credentials():
+    """Gets credentials and project ID from the GCE Metadata Service."""
+    # Ping requires a transport, but we want application default credentials
+    # to require no arguments. So, we'll use the _http_client transport which
```

Comment:
I like the idea, though you should **allow** callers to send a `Request` if they please. Users may want this, for example, to have **fully**-audited HTTP activity.